### PR TITLE
[3342] API call from token provider causing ANR issue

### PR DESCRIPTION
**Pull request by kanat**
### 🎯 Goal

To avoid getting ANR during `connectUser` call

### 🛠 Implementation details

`setUser` is being invoked in the `IO` thread to invoke `TokenProvier.loadToken` off the `Main` thread.

### 🧪 Testing

- Put `Thread.sleep(10_000L)` into `ConstantTokenProvider.loadToken`
- build & launch any Sample app
- select any user
- make sure there is no `ANR` happening

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [ ] ~~Changelog is updated with client-facing changes~~
- [ ] ~~New code is covered by unit tests~~
- [ ] ~~Comparison screenshots added for visual changes~~
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_

### CODE REVIEW

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 1]
```diff
@@ -291,25 +296,39 @@ public class ChatClient internal constructor(
 
             override fun enqueue(callback: Call.Callback<ConnectionData>) {
```
Comment (adasiewiczr) [inline stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:297]: Can't we just use `CoroutineCall` here?

Comment (kanat) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:297]: Hey @adasiewiczr 

Thanks for your feedback.
I didn't notice we have something like this 🙈

Yes we def can use `CoroutineCall` here 👍

File: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt [hunk 2]
```diff
@@ -283,30 +284,29 @@ public class ChatClient internal constructor(
...
-                        override fun onError(error: ChatError) {
-                            callback.onResult(Result(error))
-                        }
+    /**
+     * Awaits [InitConnectionListener] being invoked from either [setUser] or [setGuestUser] or [setAnonymousUser].
+     */
+    private suspend fun ((initListener: InitConnectionListener) -> Unit).awaitResult(): Result<ConnectionData> {
+        return suspendCoroutine { continuation ->
+            invoke(
+                object : InitConnectionListener() {
```
[the remark this answers is not in the document]
Comment (kanat) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:298]: Hm, I like the idea, however, we still keep the reference to that listener inside `ChatClient`.
```kotlin
private var connectionListener: InitConnectionListener? = null
```
which is getting invoked from `callConnectionListener`



Can we probably fix the ANR first and create another ticket to refactor this part?

[the remark this answers is not in the document]
Comment (kanat) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:298]: Do you want to replace `connectionListener ` with some sort of `Flow`?
Cause I assume we still need to mediator to be noticed when a connection happens.

[the remark this answers is not in the document]
Comment (JcMinarro) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:298]: Sure, I don't want to block the PR, Let's create a new ticket to investigate how to improve it.
I am not sure we are using this callback properly and I think we can remove it

[the remark this answers is not in the document]
Comment (kanat) [reply stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt:298]: Created a separate ticket: https://github.com/GetStream/stream-chat-android/issues/3364

**Review by adasiewiczr — COMMENTED** [review]
Looks great. Let's update CHANGELOG and we can merge it

**Review by adasiewiczr — APPROVED** [review]
### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] Bugs validated (bugfixes)
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

**Pull request merged**